### PR TITLE
drivers: cache: nrf: Fixes and optimization in the driver

### PR DESCRIPTION
**Pull request by nordic-krch**
Fixes:
- Hanging when performing full cache operations
- Lack of data cache flush before disabling

Optimization:
- Remove busy wait from status polling
- Remove spin lock from line operations. Using LINEADDR readout to detect preemption

Additionally, use Kconfig DCACHE line size instead of fixed value in the driver.

### CODE REVIEW

File: drivers/cache/cache_nrf.c [hunk 1]
```diff
@@ -17,7 +17,9 @@ LOG_MODULE_REGISTER(cache_nrfx, CONFIG_CACHE_LOG_LEVEL);
 #define CACHE_LINE_SIZE		     32
 #define CACHE_BUSY_RETRY_INTERVAL_US 10
 
-static struct k_spinlock lock;
+/* Some SOCs has a bug that requires to set 28th bit in the address on Trustzone secure builds. */
+#define ADDR_PATCH COND_CODE_1(CONFIG_CACHE_NRF_PATCH_LINEADDR, \
+		(COND_CODE_1(CONFIG_TRUSTED_EXECUTION_NONSECURE, (0), (BIT(28)))), (0))
```
Comment (anangl) [inline drivers/cache/cache_nrf.c:21]: Wouldn't it be more readable to just have in that affected function:
```
	if (IS_ENABLED(CONFIG_CACHE_NRF_PATCH_LINEADDR) &&
	    !IS_ENABLED(CONFIG_TRUSTED_EXECUTION_NONSECURE)) {
		addr |= BIT(28);
	}
```
and the above comment.

Comment (nordic-krch) [reply drivers/cache/cache_nrf.c]: done

File: drivers/cache/cache_nrf.c [hunk 2]
```diff
@@ -17,7 +17,9 @@ LOG_MODULE_REGISTER(cache_nrfx, CONFIG_CACHE_LOG_LEVEL);
 #define CACHE_LINE_SIZE		     32
 #define CACHE_BUSY_RETRY_INTERVAL_US 10
 
-static struct k_spinlock lock;
```
[the remark this answers is not in the document]
Comment (nordic-krch) [reply drivers/cache/cache_nrf.c:20]: done

File: drivers/cache/cache_nrf.c [hunk 3]
```diff
@@ -102,66 +91,66 @@ static inline int _cache_all(NRF_CACHE_Type *cache, enum k_nrf_cache_op op)
...
 			       size_t size)
 {
+	/* Some SOCs has a bug that requires to set 28th bit in the address on
+	 * Trustzone secure builds.
+	 */
+	if (IS_ENABLED(CONFIG_CACHE_NRF_PATCH_LINEADDR) &&
+	    !IS_ENABLED(CONFIG_TRUSTED_EXECUTION_NONSECURE)) {
+		addr |= BIT(28);
+	}
+					}
```
Comment (KAGA164) [inline drivers/cache/cache_nrf.c:138]: Unnecessary  }?

File: drivers/cache/cache_nrf.c [hunk 4]
```diff
@@ -102,66 +91,66 @@ static inline int _cache_all(NRF_CACHE_Type *cache, enum k_nrf_cache_op op)
...
 
 static inline int _cache_range(NRF_CACHE_Type *cache, enum k_nrf_cache_op op, void *addr,
 			       size_t size)
 {
+	/* Some SOCs has a bug that requires to set 28th bit in the address on
+	 * Trustzone secure builds.
+	 */
+	if (IS_ENABLED(CONFIG_CACHE_NRF_PATCH_LINEADDR) &&
+	    !IS_ENABLED(CONFIG_TRUSTED_EXECUTION_NONSECURE)) {
+		addr |= BIT(28);
```
Comment (KAGA164) [inline drivers/cache/cache_nrf.c:136]: Possible compilation error

**Review by KAGA164 — CHANGES_REQUESTED** [review]

**Comment by KAGA164** [comment]
Compilation errors after disabling CONFIG_DCACHE

File: drivers/cache/cache_nrf.c (hunk 3 above)
Comment (nordic-krch) [reply drivers/cache/cache_nrf.c]: fixed now

File: drivers/cache/cache_nrf.c (hunk 4 above)
Comment (nordic-krch) [reply drivers/cache/cache_nrf.c]: fixed.

**Review by KAGA164 — APPROVED** [review]

**Review by anangl — DISMISSED** [review]

**Review by KAGA164 — COMMENTED** [review]
Observing this issue when CONFIG_DCACHE=n:
/home/kaga/ncs/zephyr/drivers/cache/cache_nrf.c:189:23: warning: implicit declaration of function 'cache_data_flush_all'; did you mean 'cache_instr_flush_all'? [-Wimplicit-function-declaration]
  189 |                 (void)cache_data_flush_all();
      |                       ^~~~~~~~~~~~~~~~~~~~
      |                       cache_instr_flush_all

**Review by carlescufi — DISMISSED** [review]

**Comment by nordic-krch** [comment]
> Observing this issue when CONFIG_DCACHE=n:

Fixed @KAGA164 finding.

**Review by anangl — APPROVED** [review]

**Review by KAGA164 — APPROVED** [review]

File: drivers/cache/cache_nrf.c [hunk 5]
```diff
@@ -68,14 +68,6 @@ static inline int _cache_all(NRF_CACHE_Type *cache, enum k_nrf_cache_op op)
 		return -ENOTSUP;
 	}
 
-	k_spinlock_key_t key = k_spin_lock(&lock);
-
-	/*
-	 * Invalidating the whole cache is dangerous. For good measure
-	 * disable the cache.
-	 */
-	nrf_cache_disable(cache);
```
Comment (carlocaione) [inline drivers/cache/cache_nrf.c:77]: I understand the issue but invalidating the cache without disabling or flushing it means 100% crashing the kernel. Is this really something you want to expose to the user?

Comment (nordic-krch) [reply drivers/cache/cache_nrf.c:77]: invalidating of whole cache is not supported https://github.com/zephyrproject-rtos/zephyr/blob/8d762a0d2e1c9d0fa8eeabfe9f09cddf26754d37/drivers/cache/cache_nrf.c#L64

**Review by carlocaione — APPROVED** [review]

**Pull request merged**